File: webtty.js
/***********************************************************************

  webtty.js - client side script used to display output from server and 
              to send keypresses to server

  This file is part of the webtty package. The webtty package was
  written by Martin Steen Nielsen

  Development is hosted by testape.com at www.testape.com/webtty

  webtty is free software; you can redistribute it and/or modify it 
  under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  webtty is distributed in the hope that it will be useful,but 
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  
  See the GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with webtty; if not, write to the Free Software Foundation, 
  Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

***********************************************************************/



/***********************************************************************
   Resize terminal
 ***********************************************************************/

function resize(id)
    {
    var elem = document.getElementById(id);
    elem.cols+=20;
    elem.rows+=6;
    if (elem.cols>20*4) {
        elem.cols-=(20*4); elem.rows-=(6*4); }
    elem.scrollTop=elem.scrollHeight;
    return false;
    }


/***********************************************************************
   Toggles scrollbars on terminal
 ***********************************************************************/
function bars(id)
    {
    var elem = document.getElementById(id);
    if (elem.style.overflow!='scroll')
        elem.style.overflow='scroll';
    else
        elem.style.overflow='hidden';
    return false;
    }


/***********************************************************************
   closes terminal
 ***********************************************************************/
function wclose(id)
    {
    var elem = document.getElementById(id);
    elem.style.display='none';
    return false;
    }




/***********************************************************************
   constructor webtty is called from client. Constructor expects to 
   receive id of element that will hold the html for the new process. 
   Function exit_func(err) is optional, but if given, it will be called 
   on errors or if the server terminates the process early.

   Sample usage -->

   <html>
     <script type="text/javascript" src="webtty.js">
     <body onLoad="new webtty('webtty_element');">
       <div id="webtty_element"></div>
     </body>
   </html>


 ***********************************************************************/

function webtty(receiving_element_id, exit_func)
    {

    /***********************************************************************
       Creates XMLHTTPRequest object
     ***********************************************************************/
    this.create_transfer_object = function()
       {
       try { return new ActiveXObject('Msxml2.XMLHTTP');    } catch(e) {}
       try { return new ActiveXObject('Microsoft.XMLHTTP'); } catch(e) {}
       try { return new XMLHttpRequest();                   } catch(e) {}
       return null;
       }

    /***********************************************************************
       callback function whenever html data is ready
     ***********************************************************************/
    this.get_html_state = function()
        {
        if (this.xhr_out.readyState == 4)
            {
            if ((this.xhr_out.status!=200)&&(this.xhr_out.status!=0))
                this.error();
            this.element.innerHTML = this.xhr_out.responseText;
            this.session_id  = document.getElementById('webtty_session_id').innerHTML;
            document.getElementById('webtty_session_id').id='default';
            this.output_area = document.getElementById('webtty_output');
            this.output_area.id = 'default';
            var obj = this;
	    this.output_area.onkeypress = function(e) { if (obj.running) return obj.key_event(e); }
            this.output_area.onkeydown  = function(e) { if (obj.running) return obj.special_key_event(e); }
	    this.get_data();
            }
        }

    /***********************************************************************
       Request a new process and HTML data from server
     ***********************************************************************/
    this.get_html = function()
        {
        var obj = this;
        if (null == this.xhr_out)
          this.xhr_out     = this.create_transfer_object();
        this.xhr_out.onreadystatechange = function() {};
        this.xhr_out.open('GET', 'webtty.php?shell=virt', true);
        this.xhr_out.onreadystatechange = function() { if (obj.running) obj.get_html_state(); }
        this.xhr_out.send('');
        }




    /***********************************************************************
       callback function whenever process data is ready
     ***********************************************************************/
    this.get_data_state = function()
        {
	if (this.xhr_out.readyState == 4)
            {
            if ((this.xhr_out.status!=200)&&(this.xhr_out.status!=0))
                this.error();
            var data = this.xhr_out.responseText; 
            if (data == '_C_L_O_S_E_D_')
                {
                this.session_id=null;
                if (undefined != this.webtty_exit)
                    this.webtty_exit(0);
                this.cleanup();
                }
            else 
               {
               if (data != '')
                  this.handle_data(data);
	       this.get_data();
               }
            }
        }


    /***********************************************************************
       Request new data from the server
     ***********************************************************************/
    this.get_data = function()
        {
        if (null == this.session_id)
            return this.error();
        if (null == this.xhr_out)
          this.xhr_out     = this.create_transfer_object();
        var obj = this;
        this.xhr_out.onreadystatechange = function() { }
        this.xhr_out.open('GET', 'webtty.php?outp=1&id='+this.session_id, true);
        this.xhr_out.onreadystatechange = function() { return obj.get_data_state(); }
        this.xhr_out.send('');
        }


    /***********************************************************************
       exit page. Remove handles and terminate session
     ***********************************************************************/
    this.cleanup = function()
        {
        this.running     = false;
	if (this.xhr_kbd)
            {
            this.xhr_kbd.onreadystatechange = function() {};
            this.xhr_kbd.abort();
            delete this.xhr_kbd;
            }
        if (this.xhr_out)
            {
            this.xhr_out.onreadystatechange = function() {};
            this.xhr_out.abort();
            delete xhr_out;
            }
        if (null != this.session_id)
            {
            var exit_req = this.create_transfer_object();
            exit_req.open('GET', 'webtty.php?id='+this.session_id+'&shell=kill', true);
            exit_req.send('');
            delete exit_req;
  	    this.session_id=null;
            }
        }
    
    /***********************************************************************
       handles process data
     ***********************************************************************/
    this.handle_data = function(data)
        {
        this.output_area.value = this.handle_special_chars(this.output_area.value + data, 2*data.length);
        this.output_area.scrollTop=this.output_area.scrollHeight;
        }

    /***********************************************************************
       Handle special characters. Simulates applying backspace, bell etc to 
       a string
     ***********************************************************************/
    this.handle_special_chars = function(streng, max_look_back)
        {
        var startpos=0;
        var write_pos=0; 
        var new_str='';
        var subs='';

        if (max_look_back>80)
            max_look_back=80;
        
        if (streng.length>max_look_back)
            startpos = streng.length-max_look_back;

        for (tl=startpos; tl<streng.length; tl++)
            {
            if (streng.charCodeAt(tl)==8)
                { write_pos--; subs = subs.substr(0,write_pos); }
            else if (streng.charCodeAt(tl)==7)
                { ; }
            else
                { write_pos++; subs += streng.charAt(tl); }
            }
        return streng.substr(0,startpos)+subs;
        }


    /***********************************************************************
       callback function whenever data has been transmitted
     ***********************************************************************/
    this.put_kbd_state = function()
        {
        if (this.xhr_kbd.readyState == 4)
            {
      	    delete this.xhr_kbd; 
	    this.put_kbd();
            }
        }
	


    /***********************************************************************
       send data to server
     ***********************************************************************/
    this.put_kbd = function()
        {
        if ('' == this.kbd_buffer)
            return;

	this.xhr_kbd=this.create_transfer_object();
        this.xhr_kbd.open('POST', "webtty.php", true);
        var obj = this;
        this.xhr_kbd.onreadystatechange = function() { if (obj.running) return obj.put_kbd_state(); }
        this.xhr_kbd.setRequestHeader("Connection", "close")
        this.xhr_kbd.setRequestHeader("Content-Type", "application/x-www-form-urlencoded")
        this.xhr_kbd.send("id="+this.session_id+"&inp="+this.kbd_buffer);
        this.kbd_buffer='';
        }


    /***********************************************************************
      buffer data and send to server
     ***********************************************************************/
    this.buffer_next_data = function(cmd)
        {
        this.kbd_buffer += cmd;
        if (undefined == this.xhr_kbd) 
            this.put_kbd();
        }



    /***********************************************************************
       handles normal keypresses
     ***********************************************************************/
    this.key_event = function(e)
        {
        if (!e) var e = window.event
        if (e.keyCode) key = e.keyCode;
        else if (e.which) key = e.which;

        if (!(window.event || e.charCode))
          return true;

        this.buffer_next_data(escape(String.fromCharCode(key)));
        return false; 
        }


    /***********************************************************************
       handles special keypresses
     ***********************************************************************/
    this.special_key_event = function (e)
        {
	if (!e) var e = window.event
        if (e.keyCode) key = e.keyCode;
        else if (e.which) key = e.which;

        if (key==9)
            { 
            this.buffer_next_data(escape("\t")); 
            return false; 
            }
        else if (key==8)
            { 
            this.buffer_next_data(escape("\010")); 
            return false; 
            }
        else if (key == 40)
            { 
            this.buffer_next_data(escape("\033[B")); 
            return false; 
            }
        else if (key == 38)
            { 
            this.buffer_next_data(escape("\033[A")); 
            return false; 
            }
        else if (key == 38)
            { 
            this.buffer_next_data(escape("\033[A")); 
            return false; 
            }
        else if (key == 13)
            { 
            this.buffer_next_data(escape("\n")); 
            return false; 
            }
        }


    /***********************************************************************
       handles errors
     ***********************************************************************/
    this.error = function()
        {
        if (undefined != this.webtty_exit)
            this.webtty_exit(1);
        if (this.element)
	   this.element.innerHTML="Error";
	this.cleanup();
        }


    this.element     = document.getElementById(receiving_element_id);
    if (null == this.element)
        return this.error();

    this.running     = true;
    this.session_id  = null;
    this.output_area = null;
    this.kbd_buffer  = '';
    this.webtty_exit = exit_func;
    }


















/***********************************************************************
   constructor webtty_admin is called from client. Constructor expects 
   to receive id of element that will hold the html for the new process. 
   Function exit_func(err) is optional, but if given, it will be called 
   on errors or if the server terminates the process early.

   Sample usage -->

   <html>
     <script type="text/javascript" src="webtty.js">
     <body onLoad="new webtty_admin('webtty_element');">
       <div id="webtty_element"></div>
     </body>
   </html>


 ***********************************************************************/

function webtty_admin(receiving_element_id, exit_func)
    {

    this.init_super = webtty;
    this.init_super(receiving_element_id, exit_func);
    this.save = '';
    delete this.xhr_kbd;
    

    /***********************************************************************
       adds new clone terminal
     ***********************************************************************/

    this.add_clone = function(data_id)
        {
        if (null == this.output_area)
            this.error(1);

        this.output_area.innerHTML +=
          ( "<span id=s"+data_id+" style='float:left;'>" +
            "  &nbsp;&nbsp;" +
	    "  <a id=close"+data_id+" href='#' onClick='return wclose(\"s" + data_id + "\");'>close</a>&nbsp;&nbsp;" +
            "  <a id=size" +data_id+" href='#' onClick='return resize(\"t" + data_id + "\");'>size</a>&nbsp;&nbsp;"  +
            "  <a id=bars" +data_id+" href='#' onClick='return bars  (\"t" + data_id + "\");'>bars</a>"  +
            "  " +
            "  <br>&nbsp;&nbsp;<textarea rows=12 cols=40 class=webtty_clone id='t" + data_id + "'></textarea>" +
            "</span>" );

        return document.getElementById('t'+data_id);
        }


    /***********************************************************************
       Request a new process and HTML data from server - 
       replaces superclass function
     ***********************************************************************/
    this.get_html = function()
        {
        var obj = this;
        if (null == this.xhr_out)
          this.xhr_out     = this.create_transfer_object();
        this.xhr_out.open('GET', 'webtty.php?shell=admin', true);
        this.xhr_out.onreadystatechange = function() { obj.get_html_state(); }
        this.xhr_out.send('');
        }

    /***********************************************************************
       handles interpretation of admin data from admin process - 
       replaces superclass function
     ***********************************************************************/
    this.handle_data = function(data)
        {
	var nxt = 0;
        var elem;
        if ((data.charCodeAt(5)==58) && (data.charCodeAt(10)==58) )
	  this.save='';
	this.save+=data;
        while (1+nxt<this.save.length)
            {
            if ((this.save.charCodeAt(5)==58) && (this.save.charCodeAt(10)==58) )
                {
                var data_id   = this.save.substr(nxt,5);
                var data_size = 1*this.save.substr(nxt+6,4);
                var data_text = this.save.substr(nxt+11,data_size);

                if ( nxt+11+data_size > this.save.length)
                    { this.save=this.save.substr(nxt); break;}
                elem = document.getElementById('t'+data_id);
                if (elem == null)
                    elem = this.add_clone(data_id);
       
                elem.textContent += data_text;
                elem.value += data_text;

                elem.scrollTop=elem.scrollHeight;
 
                nxt += (11+data_size);
                }
            else
                nxt == this.save.length;
            }
        if (nxt == this.save.length)
            this.save='';
        }

    /***********************************************************************
       handles normal keypresses - replaces superclass function
     ***********************************************************************/
    this.key_event = function(e)
        {
        return true;
        }


    /***********************************************************************
       handles special keypresses - replaces superclass function
     ***********************************************************************/
    this.special_key_event = function (e)
        {
        return true;
        }

    }

File: webtty.c
/***********************************************************************

webtty.c - using fifos to control applications running in a pseudo
terminal.

This file is part of the webtty package. Latest version can be found
at http://testape.com/webtty_sample.html. The webtty package was 
written by Martin Steen Nielsen. 

webtty is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

webtty is distributed in the hope that it will be useful,but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with webtty; if not, write to the Free Software Foundation,
Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

***********************************************************************/

#include <unistd.h>
#include <sys/types.h>
#include <sys/time.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <pty.h>
#include <unistd.h>
#include <fcntl.h>
#include <signal.h>
#include <errno.h>
#include <time.h>
#include <syslog.h>

#define LID LOG_MAKEPRI(LOG_USER, LOG_DEBUG)

#define MAX_CHUNK 1000

//#define DEBUG syslog
#define DEBUG (void)

/***********************************************************************
wait 
***********************************************************************/
static int waitms(int ms)
    {
    struct timespec requested = { 0, ms*1e6  };
    struct timespec remaining;

    nanosleep(&requested, &remaining);

    return ms;
    }
    

/***********************************************************************
signal handler ( from GNU C Library documentation)
***********************************************************************/

static volatile sig_atomic_t fatal_error_in_progress = 0;
static int terminal, pid_terminal, pid_buffer, pid_fifo_out, pid_fifo_in;
static char *name_buffer, *name_fifo_in, *name_fifo_out, 
            *name_pid, *process_id, *name_admin;

static void fatal_error_signal (int sig)
    {
    /* Since this handler is established for more than one kind of signal,
    it might still get invoked recursively by delivery of some other kind
    of signal.  Use a static variable to keep track of that. */
    if (fatal_error_in_progress)
        raise (sig);
    fatal_error_in_progress = 1;

    /* allow client 5 second to fetch whatever 
       is there, before terminating */
    DEBUG(LID,"terminating in 5 sec");
    sleep(5);
    
    /* kill children */
    if (pid_terminal)
       {
       DEBUG(LID,"killing terminal %d",pid_terminal);
       kill(pid_terminal,sig);
       }
       
    if (pid_buffer)
       {
        DEBUG(LID,"killing buffer %d",pid_buffer);
        kill(pid_buffer,sig);
	}

    if (pid_fifo_out)
       {
        DEBUG(LID,"killing out %d",pid_fifo_out);
        kill(pid_fifo_out,sig);
	}

    if (pid_fifo_in)
       {
        DEBUG(LID,"killing in %d",pid_fifo_in);
        kill(pid_fifo_in,sig);
	}

    /* cleanup files */
    DEBUG(LID,"clean up");
    unlink(name_buffer);
    unlink(name_fifo_in);
    unlink(name_fifo_out);
    unlink(name_pid);

    /* cleanup memory */
    free(name_buffer);
    free(name_fifo_in);
    free(name_fifo_out);
    free(name_pid);
    free(name_admin);

    closelog();
    free(process_id);

    /* Now reraise the signal.  We reactivate the signal's
    default handling, which is to terminate the process.
    We could just call exit or abort,
    but reraising the signal sets the return status
    from the process correctly. */
    signal (sig, SIG_DFL);
    raise (sig);
    }

/***********************************************************************
exit error function
***********************************************************************/
static void exit_error(char *error_text)
    {
    fprintf(stderr, "failed - %s\n",error_text);
    DEBUG( LID, "failed - %s",error_text);
    _exit(1);
    }



/***********************************************************************
wait for input function ( from GNU C library documentation )
***********************************************************************/

static int input_timeout (int filedes, unsigned int milli_seconds)
    {
    char buf;
    fd_set set;
    struct timeval timeout;
    int res,r;

   /* Initialize the file descriptor set. */
    FD_ZERO (&set);
    FD_SET (filedes, &set);

    /* Initialize the timeout data structure. */
    timeout.tv_sec = 0;
    timeout.tv_usec = 1000 * milli_seconds;

    /* select returns 0 if timeout, 1 if input available, -1 if error. */
    return select (FD_SETSIZE, &set, NULL, NULL, &timeout);
    }



/***********************************************************************
handle terminal
***********************************************************************/

static int handle_terminal (int *terminal, char **argv)
    {
    int pid = forkpty (terminal, 0, 0, 0);

    if ( -1 == pid )
        /* forkpty failed */
        exit_error("could not fork or more pseudo terminals available");

    /* are we parent ? */
    if (  0 != pid )
        /* yes - return with child pid */
        return pid;

    /* remove cr/lf translation from terminal */
    struct termios settings;
    tcgetattr (STDOUT_FILENO, &settings);
    settings.c_oflag &= ~OPOST;
    settings.c_oflag &= ~ONLCR;
    tcsetattr (STDOUT_FILENO, TCSANOW, &settings);

    /* stdin/stdout/stderr is connected to fd. Launch application */
    execvp(argv[0],&argv[0]);

    DEBUG( LID, "Could not execute\n");
    /* application terminated - die and return */
    close(*terminal);
    _exit (1);
    }



/***********************************************************************
handle output buffer - copies from terminal to file buffer
***********************************************************************/

static int handle_output_buffer (int terminal_out, char *buffer)
    {
    int pid, terminal_buffer = open(buffer,O_WRONLY|O_CREAT,0777);

    if ( -1 == terminal_buffer )
        /* failed */
        exit_error("creating buffer <id>_out failed");

    pid = fork();

    if ( -1 == pid )
        /* failed */
        exit_error("fork failed");

    /* are we parent ? */
    if (  0 != pid )
        /* yes - return with child pid */
        return pid;

    /* copy from pseudo terminal to write buffer.
    Exit when ternminal dies or when killed by parent */
    while (-1 != input_timeout(terminal_out,0))
        {
        char buf;

        if ( 1 == input_timeout(terminal_out,100) )
            {
            int res = read(terminal_out,&buf,1);

            /* Reading from pseudo terminal in loop until closed */
            if ( -1 == res)
                break;
            if ( 0 ==res )
                waitms(100);
            else
	       /* keep writing to pseudo terminal */
               write(terminal_buffer,&buf,1);
            }
        }
    /* close file */
    close(terminal_buffer);
    /* terminate process */
    _exit(1);
    }


/***********************************************************************
handle output fifo
***********************************************************************/

static int handle_output_fifo (char *name_buffer, char *name_fifo_out)
    {
    int data_count, time_count, sres, fifo_out, admin_out, pid, buffer;

    buffer = open(name_buffer,O_RDONLY);
    if ( -1 == buffer )
        /* failed */
        exit_error("open buffer <id>_out failed");

    pid = fork();

    if ( -1 == pid )
        /* failed */
        exit_error("fork failed");

    /* are we parent ? */
    if (  0 != pid )
        /* yes - return with child pid */
        return pid;

    /* no - We're the output handling child. */
    while(1)
        {
        char *outbuf;
	int time=0;

        /* Open output fifo - will block until connected */
        DEBUG( LID, "output: waiting for connection");

        do
	    {
	    /* start with pause, to allow close operation to recover, before
	       reopening */
	    time += waitms(100);

	    /* Open in non - blocking will return -1 if not connected */
	    fifo_out = open(name_fifo_out,O_WRONLY|O_NONBLOCK);

	    /* Have we been here for 10 sec ? */
	    if (time>10000)
                {
                /* Yes - give up and terminate */
                DEBUG( LID, "output: giving up waiting");
                close(buffer);
                _exit(1);
                }
	    /* Keep going until client is connected */  
            } while( -1 == fifo_out);

        DEBUG( LID, "output: connected\n");

        /* zero counters */
        data_count = 0; time_count = 0;

        outbuf = (char*)malloc(MAX_CHUNK);
        while(1)
            {
            DEBUG(LID, "output: time is %d ms, data is %d bytes",time_count, data_count);

            if (1 != read(buffer,&outbuf[data_count],1))
                {
                /* Yes - simply make 50 ms pause */
                /* consider feeding browser 0's to keep connection */
                time_count += waitms(50);
                }
            else
                {
                if (data_count==0) 
		    time_count=0;
		data_count++;
                }

            /* give the browser manageable chunks */
            if (data_count>=MAX_CHUNK)
                {
                DEBUG(LID, "output: max data reached");
                break;
                }
            /* avoid timeout in browser, if no data */
            if (time_count>10000)
                {
                DEBUG(LID, "output: max time reached");
                break;
                }

            /* Transmit data after 200 ms */
            if (data_count && time_count)
                {
                DEBUG(LID, "output: timeout after %d ms (%d bytes)",time_count, data_count);
                break;
                }
            }

        if (data_count != 0 )
	  {
	  /* write data to fifo */
          DEBUG(LID, "output: writing %d bytes",data_count);
          write(fifo_out,outbuf,data_count);

  	  /* make a blocking lock request on admin file */
          admin_out = open(name_admin, O_WRONLY|O_CREAT|O_APPEND, -1);
 	  struct flock fl = { F_WRLCK };
	  fcntl (admin_out, F_SETLKW, &fl);
	
          /* duplicate data to admin file */
          char size_buf[6];
 	  sprintf(size_buf,"%.5d",pid_terminal);
  	  write(admin_out,size_buf,5);
	  write(admin_out,":",1);

 	  sprintf(size_buf,"%.4d",data_count);
	  write(admin_out,size_buf,4);
	  write(admin_out,":",1);
          write(admin_out,outbuf,data_count);

          /* Free and close fifo and adminstrative file */
  	  struct flock fu = { F_UNLCK };
	  fcntl (admin_out, F_SETLKW, &fl);
	  close(admin_out);
	  }
        free(outbuf);
	close (fifo_out);
        }

    close(buffer);
    _exit(1);
    }



/***********************************************************************
handle input fifo
***********************************************************************/

static int handle_input_fifo (char *name_fifo_in, int terminal)
    {
    int pid = fork();

    if ( -1 == pid )
        /* failed */
        exit_error("fork failed");

    /* are we parent ? */
    if (  0 != pid )
        /* yes - return with child pid */
        return pid;

    /* no - we're child. loop on open and read pipe. Exit when killed
    by parent only or when pipe disappears */
    while(1)
        {
        char buf;
        DEBUG(LID, "input: waiting");
        /* Open input pipe */
        int fifo_in = open(name_fifo_in,O_RDONLY,0);
        DEBUG(LID, "input: connected");
        /* error ? */
        if (fifo_in == -1)
            /* Yes - die */
            exit_error("input pipe not found");

        /* No - Reading from pipe in loop until closed */
        while (1 == read(fifo_in,&buf,1))
            /* keep writing to pseudo terminal */
            write(terminal,&buf,1);
        /* close this end of pipe */
        close(fifo_in);
        }
    _exit(1);
    }



/***********************************************************************
   exit help function
 ***********************************************************************/

static void exit_help(void)
    {
    /* Display the help */
    printf("This program will launch a command in a pseudo terminal. Data for the\n");
    printf("pseudo terminal can be controlled using fifo's.\n\n");
            
    printf("Usage:\n");
    printf("  webtty <id> <admin_file> <application> [parameters]\n\n");

    printf("Input to process in pseudo terminal can be sent writing to fifo <id>_in.\n");
    printf("Different controlling processes can");
    printf("share the fifo.\n The application will re-attach to the fifo input file every \n");
    printf("time it is closed.\n\n");
    printf("Output from process in pseudo terminal can be read from fifo <id>_out.\n");
    printf("Output fifo will never deliver more than 1000 bytes at a time. The controlling\n");
    printf("process can reattach several times to the fifo to get more data. If there is\n");
    printf("no data, the process will block for max 5 seconds or until data is available.\n\n");

    printf("Example:\n");
    printf("  webtty key1234 admin_log bash\n");
    printf("  webtty key1234 admin_log traceroute testape.com\n");
    exit(1);
    }



/***********************************************************************
process arguments and control process and pipe connetions
***********************************************************************/

int main(int argc, char **argv)
    {
    printf("webtty - monitor and control tty applications using fifos\n\n");

    /* Is there no args or is '--help' '-?' or '-h' present ? */
    if ((1 == argc) || strstr(argv[1],"--help-?-h"))
        /* Yes - display the help */
        exit_help();
    else
        {
        FILE *f;
	int dying_status, dying_pid;
	
        /* No - run the program */
        name_buffer   = (char *)malloc(5+strlen(argv[1]));
        name_fifo_in  = (char *)malloc(5+strlen(argv[1]));
        name_fifo_out = (char *)malloc(5+strlen(argv[1]));
        name_pid      = (char *)malloc(5+strlen(argv[1]));
        name_admin    = (char *)malloc(1+strlen(argv[2]));
        process_id    = (char *)malloc(9+1+strlen(argv[1]));
	
        strcpy(name_admin, argv[2]);
	
	sprintf(process_id, "webtty[%s]",argv[1]);
        openlog(process_id,LOG_PERROR,0);	

        /* launch application */
        pid_terminal = handle_terminal(&terminal, &argv[3]);
        
        /* handover to main process id to php */
        strcpy(name_pid, argv[1]);
        strcat(name_pid, "_pid");
        f = fopen(name_pid,"w");
        fprintf(f,"%d\n",pid_terminal);
        fclose(f);

        /* launch buffer handling */
        strcpy(name_buffer,argv[1]);
        strcat(name_buffer,"_buf");
        pid_buffer = handle_output_buffer(terminal,name_buffer);

        /* launch fifo output handling */
        strcpy(name_fifo_out, argv[1]);
        strcat(name_fifo_out, "_out");
        mkfifo(name_fifo_out,0777);
        pid_fifo_out = handle_output_fifo(name_buffer, name_fifo_out);

        /* launch fifo input handling */
        strcpy(name_fifo_in, argv[1]);
        strcat(name_fifo_in, "_in");
        mkfifo(name_fifo_in,0777);
        pid_fifo_in = handle_input_fifo(name_fifo_in,terminal);

        DEBUG(LID, "new process t:%d b:%d i:%d o:%d",pid_terminal,pid_buffer,pid_fifo_in,pid_fifo_out);
        /* establish signal handler */
        if (signal (SIGINT, fatal_error_signal) == SIG_IGN)
            signal (SIGINT, SIG_IGN);
        if (signal (SIGHUP, fatal_error_signal) == SIG_IGN)
            signal (SIGHUP, SIG_IGN);
        if (signal (SIGTERM, fatal_error_signal) == SIG_IGN)
            signal (SIGTERM, SIG_IGN);
        /* wait for any children to die */
        int dying_id =  waitpid (-1, 0, 0);
	DEBUG(LID, "dying process %d",dying_id);
	if (pid_fifo_in == dying_id)
	  pid_fifo_in=0;
	if (pid_fifo_out == dying_id)
	  pid_fifo_out=0;
	if (pid_buffer == dying_id)
	  pid_buffer=0;
	if (pid_terminal == dying_id)
	  pid_terminal=0;
	  
        /* cleanup and kill rest of children in signal handler */
	raise (SIGTERM);
	/* wait for all to die, before returning */
	wait (0);
        }
    DEBUG(LID, "ended");
    }
